Handle ApiVersions unsupported version downgrade

robobario opened the pull request:
### Type of change

- Bugfix

### Description

Closes #1611

When proxy receives an unknown ApiVersionsRequest version it:
1. does not parse any bytes beyond the latest header version known bytes.
   (making the same assumption as the broker, that future header versions
   will be backwards compatible)
2. constructs a frame containing a special v0 ApiVersionsRequest and
   forwards it on
3. intercepts that frame with an internal filter, which detects the
   message is special and then responds with an ApiVersionsResponse with
   the error code set to UNSUPPORTED_VERSION and the api versions array
   populated with the supported versions for ApiVersions.

So the broker is not exposed to this special request.

We need to handle the client sending an ApiVersionsRequest at a version
higher than the proxy understands. The broker handles this by not
decoding the message body, and responding with a v0 ApiVersionsResponse
containing an UNSUPPORTED_VERSION error code and the api_versions array
populated with the supported versions for ApiVersions. This enables the
client to retry with the highest supported ApiVersions version. Prior to
Kafka 2.5.0 it just responded with an empty api_versions array and the
client would retry with a v0 request. This means that the broker would
never see the client software name and version, which is why they
changed to sending back the supported version. See KIP-511.

This proxy behaviour means in some cases clients will now experience two
UNSUPPORTED_VERSION responses in a row, one from the proxy and then one
from the broker, if the broker again supported a lower ApiVersions
version than the proxy. From our investigations the java client and 
librdkafka support this flow.

We also considered other flows where the proxy forwards on a v0
ApiVersionsRequest to the broker and then intercepts the response to
tactically set the error code and versions. But this could cause
problematic api interactions on the server side. For example the SASL
server authenticator currently cannot tolerate two valid
ApiVersionsRequests in a row. There may be no clients affected currently
(java SASL clientsends a v0 request so is unaffected), but since we can
identify this possible bad interaction and the core clients support the extra
retry we think it's safer to put the unusual behaviour onto the clients
rather than exposing the broker to unusual sequences of messages.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Update documentation
- [ ] Make sure all unit/integration tests pass
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] If applicable to the change, [trigger the system test suite](../blob/main/DEV_GUIDE.md#jenkins-pipeline-for-system-tests).  Make sure tests pass.
- [ ] If applicable to the change, [trigger the performance test suite](../blob/main/PERFORMANCE.md#jenkins-pipeline-for-performance). Ensure that any degradations to performance numbers are understood and justified.
- [ ] Ensure the PR references relevant issue(s) so they are closed on merging.
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

> **_NOTE:_**  You must be a member of `@kroxylicious/developers` to trigger the system test and performance test suites.  If you are not part of this group, comment on the PR requesting a trigger, tagging `@kroxylicious/developers`.

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/KafkaRequestDecoder.java:
```diff
@@ -105,10 +119,24 @@ protected Frame decodeHeaderAndBody(ChannelHandlerContext ctx, ByteBuf in, final
+        requestHeaderData.setCorrelationId(correlationId);
+        requestHeaderData.setRequestApiKey(ApiKeys.API_VERSIONS.id);
+        requestHeaderData.setRequestApiVersion((short) 0);
+        ApiVersionsRequestData apiVersionsRequestData = SPECIAL_DOWNGRADED_REQUEST_DATA;
```
You are using the singleton instance of SPECIAL_DOWNGRADED_REQUEST_DATA so that the DowngradeFilter can distinguish between a client using version 0 and the downgrade case.

An alternative would be to subclass ApiVersionsRequestData as DowngradedApiVersionsRequestData and let the  filter use an `instanceof`.

robobario replied:
Good idea Keith, I have reimplemented it this way and overridden a few methods so that if anything attempted to write the special message as byte the `size` and `write` methods will throw, as an extra defence in case we lost the short-circuiting in a refactor.

k-wall commented:
I have tested in detail but I think the approach looks good.

SamBarker reviewed (approved):
I think this makes sense but I haven't tried it out locally (and wont be able to while travelling)

k-wall commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/ApiVersionsDowngradeIT.java:
```diff
@@ -0,0 +1,105 @@
+ * for ApiVersions, so that the client can retry with the highest supported ApiVersions version.
+ * </p>
+ * <br>
+ * In this case the proxy will respond with it's own v0 response populated with the proxies supported
```
```suggestion
 * In this case the proxy will respond with its own v0 response populated with the proxies supported
```

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/KafkaRequestDecoder.java:
```diff
@@ -21,9 +22,12 @@
 public class KafkaRequestDecoder extends KafkaMessageDecoder {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(KafkaRequestDecoder.class);
+    public static final ApiVersionsRequestData SPECIAL_DOWNGRADED_REQUEST_DATA = new ApiVersionsRequestData();
```
```suggestion
```
dead code.

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/KafkaRequestDecoder.java:
```diff
@@ -85,6 +88,17 @@ protected Frame decodeHeaderAndBody(ChannelHandlerContext ctx, ByteBuf in, final
         }
         final RequestFrame frame;
         if (decodeRequest) {
+            short highestProxyVersion = apiKey.latestVersion(true);
```
I wondered about `enableUnstableLastVersion = true`.  I wondered if this could mean a production environment could start using unstable versions spontaneously.  I decided that `true` is okay.  The client and broker would need to also have the same configuration for this is to be true. 

I think what you have done is okay.

robobario replied:
here's the commit where it was [added](https://github.com/kroxylicious/kroxylicious/pull/389/commits/7f2e4e51eac41a0fe645cc2e59974398432b0066)

I feel like I discussed it somewhere, but can't dig it up. The commit says:

> 3. `ApiMessageType#highestSupportedVersion()` has been removed/replaced
with `ApiMessageType$highestSupportedVersion(boolean enableUnstableLastVersion)`
I've made Kroxylicious support the unstable last version as it should be the
decision of the backing broker whether to accept unstable versions, the
proxy should handle them.

So yeah, the thought was by default the proxy shouldn't interfere if the client/broker want to use unstable versions.

k-wall commented:
@robobario I think your approach is good.  I wanted to get some confidence of some tests against a real client https://github.com/robobario/kroxylicious/pull/15.  WDYT?   I also still want to do some testing against a librdkafka based client, I didn't quite reach that today.

robobario commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/ApiVersionsDowngradeIT.java:
```diff
@@ -0,0 +1,105 @@
+        short requestHeaderVersion = ApiKeys.API_VERSIONS.requestHeaderVersion(unsupportedVersion);
+        ObjectSerializationCache cache = new ObjectSerializationCache();
+        int headerSize = requestHeaderData.size(cache, requestHeaderVersion);
+        int bodySize = 4;
```
thanks i was searching around for that, I know I've seen it in Kroxy code before

robobario replied:
I changed to `byte[]` instead so we only need the one var and can use `length`

robobario commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/codec/DowngradeApiVersionsRequestData.java:
```diff
@@ -0,0 +1,44 @@
+ * This subclass is used when we receive an ApiVersions request at an api version higher
+ * than the proxy supports. It should be handled internally with a short-circuit response
+ * and not forwarded to any broker.
+ */
```
good point, I'll make them private classes of ApiVersionsDowngradeFilter to hide them away further

robobario commented:
Thanks @k-wall I've incorporated https://github.com/robobario/kroxylicious/pull/15. Some differences:

1. I've made it the `ApiVersionsServiceImpl`'s responsibility to know the latest supported version
2. I also made it sanity check at construction time that the user hasn't configured a latest version lower than the oldest version.
3. rebased things around so that the version override mechanism is in place before making the KIP-511 changes, if this PR gets unwieldy it should be easy to extract it into it's own PR now.
4. added in an ApiVersionsIT checking the intersection logic applies the overrides

robobario commented:
@SamBarker I've gone ahead and incorporated a change from Keith which is an enabler for integration testing:

1. java client on X, proxy on X-1, broker on X

with real clients.

The user can override the max version for an ApiKey via an experimental proxy configuration, so we can pretend the proxy doesn't understand `v4` and trigger the fallback mechanism.

This lets us test that at least in this one case the java client can deal with a downgrade at the proxy (with plain and SASL)

robobario commented:
a comment from @tombentley , we should consider distinguishing these from `experimental` but user facing configurations and make sure you have to jump through some hoops to switch these features on, so that this clearly some 'for-testing-only' feature. For example by setting an env/system var like `KROXYLICIOUS_ENABLE_INTERNAL_EXTENSIONS`, or having an environment enum where we could pass `Environment.DEVELOPMENT` to `KafkaProxy`. The field might be better named `internal` then `experimental`.

k-wall reviewed (approved):
lgtm

k-wall commented:
I spent a bit more time testing this change today.  All looks good.  I tested against Redpanda (broker) and kafkajs (pure JS client).  In all cases, the api-versions downgrading seems to be good.

It'd be good to get some tests from using other clients into the testsuite, but I think this can wait.

robobario commented:
Have pushed up something so users and tests have to opt into these internal features

This adds a concept of proxy environment where we identify two current
environments:
1. PRODUCTION: a proxy configured for production, internal development
   extensions are disallowed
2. DEVELOPMENT: a proxy configured for testing, internal development
   extensions are allowed

By default the proxy selects PRODUCTION as it's environment and will
fail to start with an exception if any internal configurations are
specified in the config file.

To opt in to the development environment the user must supply either:
1. a system prop `-DKROXYLICIOUS_ENVIRONMENT=DEVELOPMENT`
2. an env var `KROXYLICIOUS_ENVIRONMENT=DEVELOPMENT`
with the system-prop taking precedence.

Integration tests behave the same way, configuring the proxy with a
PRODUCTION environment, you have to set it to DEVELOPMENT explicitly
in your test to use any internal configurations.

k-wall commented:
@robobario thanks, I'll review the new work soon.

Did you have chance to look at https://github.com/robobario/kroxylicious/pull/16?

robobario commented:
> Did you have chance to look at https://github.com/robobario/kroxylicious/pull/16?

Nope, missed that.

k-wall commented on kroxylicious-app/src/main/java/io/kroxylicious/app/Kroxylicious.java:
```diff
@@ -35,13 +35,18 @@ public class Kroxylicious implements Callable<Integer> {
     private static final Logger LOGGER = LoggerFactory.getLogger("io.kroxylicious.proxy.StartupShutdownLogger");
     private static final String UNKNOWN = "unknown";
-    private final BiFunction<PluginFactoryRegistry, Configuration, KafkaProxy> proxyBuilder;
+    public static final String KROXYLICIOUS_ENVIRONMENT = "KROXYLICIOUS_ENVIRONMENT";
```
KROXYLICIOUS_ENVIRONMENT is a bit vague.   I'd suggest DEVELOPMENT_MODE which defaults to false.

k-wall commented on kroxylicious-app/src/test/java/io/kroxylicious/app/KroxyliciousIT.java:
```diff
@@ -43,11 +51,76 @@ class KroxyliciousIT {
+        assertThat(lastProcess.onExit()).succeedsWithin(5, TimeUnit.SECONDS);
+        byte[] bytes = lastProcess.getInputStream().readAllBytes();
+        String output = new String(bytes, StandardCharsets.UTF_8);
+        assertThat(output).contains("internal configuration for proxy present in production environment");
```
To avoid introducing the term "internal" we could just say "development mode configurations present without developer mode enabled"

k-wall commented:
lgtm
